testkit-backend/src/main/java/neo4j/org/testkit/backend/messages/requests/deserializer/types/CypherDateTime.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) "Neo4j"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package neo4j.org.testkit.backend.messages.requests.deserializer.types;
+
+import java.time.DateTimeException;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import org.neo4j.driver.Value;
+import org.neo4j.driver.internal.value.DateTimeValue;
+import org.neo4j.driver.internal.value.LocalDateTimeValue;
+
+public class CypherDateTime implements CypherType {
+    private final int year, month, day, hour, minute, second, nano;
+    private final String zoneId;
+    private final Integer offset;
+
+    public CypherDateTime(
+            int year,
+            int month,
+            int day,
+            int hour,
+            int minute,
+            int second,
+            int nanosecond,
+            String zoneId,
+            Integer offset) {
+        this.year = year;
+        this.month = month;
+        this.day = day;
+        this.hour = hour;
+        this.minute = minute;
+        this.second = second;
+        this.nano = nanosecond;
+        this.zoneId = zoneId;
+        this.offset = offset;
+    }
+
+    @Override
+    public Value asValue() {
+        if (offset == null) {
+            return new LocalDateTimeValue(LocalDateTime.of(year, month, day, hour, minute, second, nano));
+        }
+        if (zoneId == null) {
+            ZoneOffset zoneOffset = ZoneOffset.ofTotalSeconds(offset);
+            return new DateTimeValue(ZonedDateTime.of(year, month, day, hour, minute, second, nano, zoneOffset));
+        }
+        try {
+            return new DateTimeValue(ZonedDateTime.of(year, month, day, hour, minute, second, nano, ZoneId.of(zoneId)));
+        } catch (DateTimeException e) {
+            return null;
```

Comment:
I have a feeling that implementation of this method should:
- check `zoneId != null` and create
- check `offset != null` and create
- create local
- throw if there is a type creation issue

Reply:
Throwing is a good point. Just a plain RuntimeException?

The order is equivalent, as TestKit guarantees that either non, only `offset`, or `offset` and `zoneId` are set. But your suggested order is easier to read and understand, I think. I'm on it.